### PR TITLE
3152 create gitlab agents project from the module

### PR DESCRIPTION
**Pull request by Stevesibilia**
___

### **PR Type**
Enhancement

___

### **Description**
- Implemented the ability to create a GitLab project automatically if `gitlab_project_name` is set
- Added new variables `gitlab_project_name` and `gitlab_root_namespace` for project creation and namespace specification
- Updated main.tf to handle both existing and newly created projects
- Modified outputs and resource references to use the new flexible project ID
- Updated CHANGELOG.md and README.md to document the new feature and version 0.7.0
- Improved error handling and conditions for project creation and referencing
- Ensured backward compatibility with existing setups

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>main.tf</strong><dd><code>Implement GitLab project creation and flexible project referencing</code></dd></summary>
<hr>

main.tf

<li>Added logic to create a new GitLab project if not existing<br> <li> Introduced variables for project creation and root namespace<br> <li> Modified resource and data source references to use the new project<br> <li> Updated GitLab agent and token configurations to use the new project <br>ID<br>

</details>

  </td>
  <td><a href="https://github.com/sparkfabrik/terraform-gitlab-kubernetes-gitlab-agent/pull/11/files#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbb">+19/-5</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>outputs.tf</strong><dd><code>Update project ID output for flexibility</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

outputs.tf

<li>Updated the <code>gitlab_agents_project_id</code> output to use the new <br><code>local.project_id</code><br>

</details>

  </td>
  <td><a href="https://github.com/sparkfabrik/terraform-gitlab-kubernetes-gitlab-agent/pull/11/files#diff-de6c47c2496bd028a84d55ab12d8a4f90174ebfb6544b8b5c7b07a7ee4f27ec7">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>variables.tf</strong><dd><code>Add new variables for project creation and root namespace</code></dd></summary>
<hr>

variables.tf

<li>Added new variables: <code>gitlab_project_name</code> and <code>gitlab_root_namespace</code><br> <li> Updated description for <code>gitlab_project_path_with_namespace</code><br>

</details>

  </td>
  <td><a href="https://github.com/sparkfabrik/terraform-gitlab-kubernetes-gitlab-agent/pull/11/files#diff-05b5a57c136b6ff596500bcbfdcff145ef6cddea2a0e86d184d9daa9a65a288e">+11/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Update CHANGELOG for version 0.7.0</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

CHANGELOG.md

<li>Added entry for version 0.7.0<br> <li> Documented new feature to create GitLab agents project<br>

</details>

  </td>
  <td><a href="https://github.com/sparkfabrik/terraform-gitlab-kubernetes-gitlab-agent/pull/11/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Update README with new project creation feature</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

README.md

<li>Added information about the new feature to create GitLab project <br>automatically<br> <li> Mentioned the requirement for proper permissions to create the project<br> <br>

</details>

  </td>
  <td><a href="https://github.com/sparkfabrik/terraform-gitlab-kubernetes-gitlab-agent/pull/11/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,14 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
 
+## [0.7.0] - 2024-10-21
+
+[Compare with previous version](https://github.com/sparkfabrik/terraform-gitlab-kubernetes-gitlab-agent/compare/0.6.0...0.7.0)
+
+### Added
+
+- The module can create the gitlab agents project by setting the variable `gitlab_project_name`
```
Comment (Monska85) [inline CHANGELOG.md:17]:
```suggestion
## [0.7.0] - 2024-10-22

[Compare with previous version](https://github.com/sparkfabrik/terraform-gitlab-kubernetes-gitlab-agent/compare/0.6.0...0.7.0)

### Added

- The module can create the gitlab agents project by setting the variable `gitlab_project_name`.
```

File: README.md [hunk 2]
```diff
@@ -8,6 +8,8 @@ If required (`gitlab_agent_grant_access_to_entire_root_namespace` configured to
 
 **ATTENTION**: you have to manually create the project that will host the Gitlab Agent configuration in Gitlab before running this module.
 
+From version 0.7.0 if you set `gitlab_project_name` the module will create gitlab project automatically. This new behavior requires the provider to have the proper permissions to create the project in the namespace.
```
Comment (Monska85) [inline README.md:11]:
```suggestion
From version `0.7.0`, if you set `gitlab_project_name` the module will create Gitlab project automatically. This new behavior requires the provider to have the proper permissions to create the project in the namespace.
```

File: main.tf [hunk 3]
```diff
@@ -20,26 +23,37 @@ locals {
...
+data "gitlab_metadata" "this" {}
+
 data "gitlab_project" "this" {
+  count               = local.use_existing_project
   path_with_namespace = var.gitlab_project_path_with_namespace
 }
 
 data "gitlab_group" "root_namespace" {
-  group_id = data.gitlab_project.this.namespace_id
+  #group_id = data.gitlab_project.this.namespace_id
```
Comment (Monska85) [inline main.tf:39]:
```suggestion
```

I would prefer to avoid the commented code.

File: variables.tf [hunk 4]
```diff
@@ -1,3 +1,14 @@
+variable "gitlab_project_name" {
+  description = "The name of the Gitlab project that hosts the Gitlab Agent configuration. If not provided, the module will use the project defined in `gitlab_project_path_with_namespace`."
+  type        = string
+  default     = ""
+}
+
+variable "gitlab_root_namespace" {
+  description = "The full path of the root namespace in Gitlab. This is the namespace where the Gitlab Agent will be granted access."
+  type        = string
+}
```
Comment (Monska85) [inline variables.tf:10]: Is it possible to make this variable optional?

I think we can use the previous logic to determine the root namespace when the `gitlab_project_name` is not empty. What do you think?

File: main.tf [hunk 5]
```diff
@@ -9,6 +9,9 @@ locals {
 
   final_namespace = var.create_namespace ? resource.kubernetes_namespace_v1.this[0].metadata[0].name : data.kubernetes_namespace_v1.this[0].metadata[0].name
 
+  use_existing_project = var.gitlab_project_name == "" ? 1 : 0
+  project_id           = local.use_existing_project == 1 ? data.gitlab_project.this[0].id : gitlab_project.project[0].id
+
```
Comment (Monska85) [inline main.tf:14]:
```suggestion
  project_id                                      = local.use_existing_project == 1 ? data.gitlab_project.this[0].id : gitlab_project.project[0].id
  project_path_with_namespace = local.use_existing_project == 1 ? data.gitlab_project.this[0].path_with_namespace : gitlab_project.project[0].path_with_namespace

```

I think we can manage conditional variable here in order to keep the code below easier to read. What do you think?

File: main.tf [hunk 6]
```diff
@@ -20,26 +23,37 @@ locals {
   # Gitlab Agent CI/CD variables
   gitlab_agent_kubernetes_context_variables = {
     (var.gitlab_agent_variable_name_agent_id) : gitlab_cluster_agent.this.name,
-    (var.gitlab_agent_variable_name_agent_project) : data.gitlab_project.this.path_with_namespace,
+    (var.gitlab_agent_variable_name_agent_project) : local.use_existing_project == 1 ? data.gitlab_project.this[0].path_with_namespace : gitlab_project.project[0].path_with_namespace,
```
Comment (Monska85) [inline main.tf:26]:
```suggestion
    (var.gitlab_agent_variable_name_agent_project) : local.project_path_with_namespace,
```

This change is dependent from the `project_path_with_namespace` local variable.

**Review by Monska85 — CHANGES_REQUESTED** [review]

File: variables.tf (hunk 4 above)
Comment (Stevesibilia) [reply variables.tf:10]: We always have the root namespace (is the key of the map we pass to the module), mapping it to a variable makes the code simplier and more readable.

The previuos logic relies on the project being already created so we can use the project name to find out the root namespace. When `gitlab_project_name` is not empty we don't have any means to find out the root namespace as the project doesn't exist yet.

**Pull request merged**